integration/compose/src/main/java/com/bumptech/glide/integration/compose/GlideModifier.kt

```diff
@@ -0,0 +1,252 @@
+      drawable.draw(drawContext.canvas.nativeCanvas)
+    }
+
+    // Allow chaining of DrawModifiers
+    drawContent()
+  }
+
+  override val shouldAutoInvalidate: Boolean
+    get() = false
+
+  override fun onAttach() {
+    super.onAttach()
+    if (currentJob == null && requestBuilder != null) {
+      currentJob = launchRequest(requestBuilder!!)
+    }
+  }
+
+  override fun onReset() {
+    super.onReset()
+    clear()
+    updateDrawable(null)
+    // onReset means we're about to updated with a new request. Unfortunately onAttach is called
+    // before we get the new request. To avoid restarting an obsolete request in onAttach, null
+    // out the request builder here.
+    requestBuilder = null
+  }
+
+  @OptIn(ExperimentGlideFlows::class, InternalGlideApi::class)
+  private fun launchRequest(requestBuilder: RequestBuilder<Drawable>) =
+    (coroutineScope + Dispatchers.Main.immediate).launch {
+      requestBuilder.flowResolvable(resolvableGlideSize).collect {
+        val (state, drawable) =
+          when (it) {
+            is Resource -> Pair(RequestState.Success(it.dataSource), it.resource)
+            is Placeholder -> {
+              val drawable = it.placeholder
+              val state = when (it.status) {
+                Status.CLEARED, Status.RUNNING -> RequestState.Loading
+                Status.FAILED -> RequestState.Failure
+                Status.SUCCEEDED -> throw IllegalStateException()
+              }
+              Pair(state, drawable)
+            }
+          }
+        updateDrawable(drawable)
+        this@GlideNode.state = state
+        invalidateDraw()
+      }
+    }
+
+  private fun updateDrawable(drawable: Drawable?) {
+    this.drawable?.callback = null
+    this.drawable = drawable
+    drawable?.colorFilter = colorFilter?.asAndroidColorFilter()
+    drawable?.callback = object : Drawable.Callback {
+      override fun invalidateDrawable(who: Drawable) {
+        invalidateDraw()
+      }
+
+      override fun scheduleDrawable(who: Drawable, what: Runnable, `when`: Long) {}
+      override fun unscheduleDrawable(who: Drawable, what: Runnable) {}
+    }
+  }
+
+  override fun onDetach() {
+    super.onDetach()
+    clear()
+  }
+
+  private fun clear() {
+    currentJob?.cancel()
+    currentJob = null
+    state = RequestState.Loading
+    updateDrawable(null)
+  }
+
+  @OptIn(InternalGlideApi::class)
+  override fun MeasureScope.measure(
+    measurable: Measurable,
+    constraints: Constraints
+  ): MeasureResult {
+    when (val currentSize = resolvableGlideSize) {
+      is AsyncGlideSize -> {
+        val inferredSize = constraints.inferredGlideSize()
+        if (inferredSize != null) {
+          currentSize.setSize(inferredSize)
+        }
+      }
+      // Do nothing.
+      is ImmediateGlideSize -> {}
+    }
+    val placeable = measurable.measure(constraints)
+    return layout(placeable.width, placeable.height) { placeable.place(0, 0) }
```

Comment:
currently you are not really influencing layout at all here, as you are just assuming the size of your children and placing at 0,0.

Ideally you would look at the drawable's intrinsic size or the size specified by resolvableGlideSize and take up that size here. However consumers are supposed to specify the size of the image they want, here is where you should use it. Then inside of draw you can just draw the image to take up the full amount of space allocated to it.

Additionally, I think if the user did not tell you what size they wanted, a sensible default would be to occupy the space given to you (constraints.maxWidth), and maintain aspect ratio.

Reply:
Yeah this is interesting. We really want to occupy the size specified by the Modifier, and ideally nothing else.

If we base the layout on the size of the image you load, then your layout is likely to shift when the image load completes, which is ugly. You could fix that by providing a specific size to Glide, but if you can do that you might as well just do it in your layout instead. Moreover you may very well want to load an image that's smaller than the layout size to save memory or reduce jank.

With all that said, can I simply rely on other modifier attributes to specify an explicit size? If I use maxWidth, will that override any sizing information provided by Modifiers earlier in the chain?

Reply:
This probably deserves a full conversation along with some examples, but.... i think there's more that you should do here than you are currently probably regardless.

I think a reasonable behavior would be:


1. if fixed constraints come in you should honor that size
2. if width is constrained and height is not, or vice versa, you should fill to that constrained width/height and utilize the aspect ratio of the image to further constrain the other

Reply:
Thanks! I've switched to using the constraint's max dimensions. I filed https://github.com/bumptech/glide/issues/5234 to follow up. I think (hope) this is pretty similar to what we had before, so ideally we can tackle more complicated layout issues in a subsequent change.